### ISSUE TITLE
Increased step time when using `transformer_engine.pytorch.Linear`

### ISSUE THREAD

**Issue by cavdard** [NVIDIA/TransformerEngine#127]
I am experiencing increased step time when using `transformer_engine.pytorch.Linear` instead of `torch.nn.Linear` with fp16 training.

Mean step time(forward+backward) using single `te.Linear` is 30% higher than `torch.nn.Linear` with the following test. Backward pass takes significantly longer.

Set up: TransformerEngine release v0.7,  Pytorch 2.0.0, A100 GPU. 

Test with single linear layer:

```
import transformer_engine.pytorch as te

hidden_size = 4096
sequence_length = 2048
batch_size = 4
dtype = torch.float16

input = torch.rand(sequence_length, batch_size, hidden_size).cuda().to(dtype=dtype)
input_y = torch.rand(sequence_length, batch_size, 3 * hidden_size).cuda().to(dtype=dtype)

linear_layer = te.Linear(
    hidden_size,
    3 * hidden_size,
    bias=True,
    params_dtype=dtype,
)

linear_layer.to(dtype=dtype).cuda()

start = torch.cuda.Event(enable_timing=True)
end = torch.cuda.Event(enable_timing=True)

number_of_iterations = 50 

start.record()
for i in range(number_of_iterations):
    with te.fp8_autocast(enabled=False):
        loss = linear_layer(input)
    loss.backward(input_y)
end.record()

torch.cuda.synchronize()

print(f"step time: {start.elapsed_time(end)/number_of_iterations}")
```

**Comment by ptrendx** [comment]
@timmoon10 Could you take a look?

**Comment by ksivaman** [comment]
TransformerEngine currently calculates activation gradients (`dgrad`) even when `input.requires_grad==False`, which is the case in your repro script. Due to this limitation in TE, the perf measurements observed above are expected since calling `transformer_engine.pytorch.Linear` executes 1 extra matrix multiply compared to `torch.nn.Linear` (3 vs 2). Other than this, the performance for both APIs is identical. We will soon have a PR to address this issue.
